core/include/scipp/core/element/rebin.h

```diff
@@ -17,6 +17,74 @@
 
 namespace scipp::core::element {
 
+struct AscendingRebin {};
+struct DescendingRebin {};
+
+namespace {
+void rebin_error() {
+  throw std::runtime_error(
+      "Unknown type of rebin operation. "
+      "Possible branches are AscendingRebin and DescendingRebin.");
+}
+} // namespace
+
+template <class RebinType, class OldType, class NewType>
+bool compare_new_high_with_old_low(const NewType xn_high,
+                                   const OldType xo_low) {
+  if constexpr (std::is_same_v<RebinType, AscendingRebin>)
+    return xn_high <= xo_low;
+  else if constexpr (std::is_same_v<RebinType, DescendingRebin>)
+    return xn_high >= xo_low;
+  else
+    rebin_error();
+}
```

Comment:
Look at how things like this are solved in the standard library, e.g., in `std::sort`: The user can pass a functor (e.g., a lambda, or `std::less`. Most of these functions should not be needed, I think. You are defining helpers for `<=` (2x) and `>` (1x). By swapping argument order you can use `std::less` (or [](const auto a, const auto b) { return a < b; }` for these 3 cases. For the decending rebin just pass `std::greater`.